Donation-platform/src/main/java/com/google/servlets/PersonalizationServlet.java

```diff
@@ -48,8 +53,25 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
       e.printStackTrace();
     }
     // Iterate over the mapping and add the tag values to the tags list
+    int numberOfBlankTags = 0;
     for (Map.Entry<String, String> entry : tagMap.entrySet()) {
-      tags.add(entry.getValue());
+      String tag = entry.getValue();
+      if(tag.length() == 0) {
+        numberOfBlankTags++;
+      }
+      tags.add(tag);
+    }
+
+    // If a user session already exists and new tags have not been selected
+    // (which is signified by the number of blank selected tags equalling 3), 
+    // then use the tags previously selected in this session.
```

Comment:
I don't quite understand this logic here of 3. Why the blank tags will be 3?

Reply:
If the user does not select any tags, there will be 3 blank tag entries (because there are 3 tag entry fields). If the user selects no new tags, then we display results based on the tags previously selected in the current session.

Reply:
numberOfBlankTags refers to the number of unselected tag fields in the current html form (not the number of blank tags in the previously-selected tags that are saved in the session).